### PR TITLE
Optimize shortsplitting

### PR DESCRIPTION
**Pull request by jridgewell**
This optimizes the short value splitting to avoid a string
allocation, taking advantage of regex's `lastIndex` instead.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -17,8 +16,9 @@ function parse(args, aliases, defaults, unknown, out) {
     var arg = args[i]
 
     if ("--" === arg) {
-      push.apply(_, args.slice(i + 1))
-      break
+      while (++i < len) {
```
Comment (jorgebucaran) [inline index.js]: @jridgewell This is genius! Good job & thanks! 🏄

Comment (jorgebucaran) [reply index.js]: And we don't need the `break` either, clever 👍

File: index.js [hunk 2]
```diff
@@ -38,10 +38,17 @@ function parse(args, aliases, defaults, unknown, out) {
           }
         }
       } else {
-        var end = arg.slice(2).search(SHORTSPLIT) + 2
-        var value = end === arg.length
-          ? (j = i + 1) === len || "-" === args[j][0] || args[(i = j)]
-          : arg.slice(end)
+        SHORTSPLIT.lastIndex = 2
```
Comment (jorgebucaran) [inline index.js:41]: @jridgewell Could we make this a separate PR please? 🙏 

Thanks.

**Review by jorgebucaran — CHANGES_REQUESTED** [review]

File: index.js (hunk 2 above)
Comment (jridgewell) [reply index.js:41]: Do you want me to split all three up?

Comment (jorgebucaran) [reply index.js:41]: Yes, please, that will help me get stuff reviewed and merged faster. Thanks.

**Comment by jridgewell** [comment]
This is now just the short splitting change.

**Comment by jridgewell** [comment]
Made this one even better:

```
getopts × 1,136,787 ops/sec
pr × 1,092,419 ops/sec

getopts × 1,129,234 ops/sec
pr × 1,185,734 ops/sec

getopts × 1,143,762 ops/sec
pr × 1,247,025 ops/sec

getopts × 1,137,478 ops/sec
pr × 1,180,274 ops/sec

getopts × 1,137,673 ops/sec
pr × 1,221,343 ops/sec
```

File: index.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-const SHORTSPLIT = /$|[!-@\[-`{-~]/
+const SHORTSPLIT = /[!-@\[-`{-~].*|$/g
```
Comment (jorgebucaran) [inline index.js]: @jridgewell What's the idea behind moving `|$` to the end and adding `g`? We only care about the first match.

Comment (jorgebucaran) [reply index.js]: @jridgewell `lastIndex`, got it! 👍 You demon! 😈

Comment (jridgewell) [reply index.js]: > What's the idea behind moving `|$` to the end

My first commit removed it, then I added it back 😅, but put it at the end (where I usually would put it). I'll move that back.

> adding `g`

There can only ever be one match, but the `g` flag allows `lastIndex` to work. Without it, `RegExp#exec` starts from the beginning every time.

**Comment by jorgebucaran** [comment]
@jridgewell I thought I got it, but it had nothing to do with `lastIndex`?

Why are you using `/g`?

**Comment by jridgewell** [comment]
See Item 8 of http://ecma-international.org/ecma-262/8.0/index.html#sec-regexpbuiltinexec. Without `g` flag, `lastIndex` gets reset to `0` (start search from beginning). That would cause our regex to always match the `-` at the beginning of `arg`.

With `lastIndex = 2` (and `g` flag so it doesn't get reset), we can start the search at the third character, saving us a `String#slice` call.

**Comment by jorgebucaran** [comment]
@jridgewell I thought regex `g` would be slow.

http://ecma-international.org/ecma-262/8.0/index.html#sec-regexpbuiltinexec 👀👍

**Comment by jridgewell** [comment]
Nah, the only difference should be:

```c++
// Notice, this is browser's C++
// So, very fast
if (!regex->global) {
  regex->lastIndex = 0;
}
```

**Comment by JorgeBucaran** [comment]
@jridgewell That works for me! You must be fun at parties! 😉 Just kidding.

So, `match[0]` holds the entire string after the SHORTSPLIT (or `""` if there was no match, on which case the previous `end === arg.length` behavior kicks in) and match.index gives us `end` for free.

Perfect.
